### PR TITLE
[WiP] Prototype of JWT with Authorization header;  much cleanup needed

### PR DESCRIPTION
**Pull request by drewwills**
Creating a PR for discussion of this change;  refinement is needed.

### CODE REVIEW

File: src/main/java/org/apereo/portlet/soffit/Headers.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apereo.portlet.soffit;
+
+public enum Headers {
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/Headers.java:24]: - [x] Move other header-defining constants, from both "sides," to this enumeration

File: src/main/java/org/apereo/portlet/soffit/connector/SoffitConnectorController.java [hunk 2]
```diff
@@ -67,6 +68,11 @@
 public class SoffitConnectorController implements ApplicationContextAware {
 
     /**
+     * Prepended to the Authorization HTTP header.
+     */
+    public static final String BEARER_PREFIX = "Bearer ";
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/connector/SoffitConnectorController.java]: - [x] Move to the Headers enum(?)

File: src/main/java/org/apereo/portlet/soffit/connector/SoffitConnectorController.java [hunk 3]
```diff
@@ -280,6 +290,18 @@ private Object buildPayload(final RenderRequest req, final RenderResponse res) {
 
     }
 
+    // TODO:  Refactor
+    private org.apereo.portlet.soffit.model.v1_0.UserDetails getUserDetails(Object payload) {
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/connector/SoffitConnectorController.java]: - [x] Make & execute a sensible, pluggable plan for how to construct JWTs

File: src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * Responsible for issuing and parsing Bearer tokens.
+ *
+ * @author drewwills
+ */
+@Service
+public class UserDetailsService {
+
+    public static final String JWT_ISSUER = "Soffit";
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java]: - [x] Will be used by more than one `@Service`;  needs to be moved to a general location

File: src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+         */
+
+        private final String name;
+
+        private Keys(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public String toString() {
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java]: - [x] Change to `getName()`

File: src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+    }
+
+    @Value("${org.apereo.portlet.soffit.model.v1_0.UserDetailsService.signatureKey}")
+    private String signatureKey;
+
+    public UserDetails createUserDetails(String username, Map<String,List<String>> attributes, List<String> groups) {
+
+System.out.println(" ## signatureKey="+signatureKey);
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java]: - [x] Remove

File: src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                .setSubject(username);
+        claims.put(Keys.ATTRIBUTES.toString(), attributes);
+        claims.put(Keys.GROUPS.toString(), groups);
+
+        final String bearerToken = Jwts.builder()
+                .setClaims(claims)
+                .signWith(SignatureAlgorithm.HS512, signatureKey)
+                .compact();
+
+        // TODO:  Encryption
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java]: - [ ] Encryption

File: src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    public UserDetails parseBearerToken(String bearerToken) {
+
+        final Jws<Claims> claims = Jwts.parser()
+                .setSigningKey(signatureKey)
+                .parseClaimsJws(bearerToken);
+
+        final String username = claims.getBody().getSubject();
+        final Map<String,List<String>> attributes = (Map<String, List<String>>) claims.getBody().get(Keys.ATTRIBUTES.toString());
+        final List<String> groups = (List<String>) claims.getBody().get(Keys.GROUPS.toString());
+
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java]: - [ ] Decryption

File: src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+
+/**
+ * Responsible for issuing and parsing Bearer tokens.
+ *
+ * @author drewwills
+ */
+@Service
+public class UserDetailsService {
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/connector/UserDetailsService.java]: - [ ] Implement some sort of token service interface

File: src/main/java/org/apereo/portlet/soffit/model/v1_0/UserDetails.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Encapsulates username, attributes, and group affiliations.
+ *
+ * @author drewwills
+ */
+public class UserDetails {
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/model/v1_0/UserDetails.java]: - [x] What _should_ be the name of this class?  `Principal`?  `User`?  `UserDetails`?

File: src/main/java/org/apereo/portlet/soffit/model/v1_0/UserDetails.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        return attributes;
+    }
+
+    public List<String> getGroups() {
+        return groups;
+    }
+
+    @Override
+    public String toString() {
+        return "UserDetails [bearerToken=" + bearerToken + ", username=" + username + ", attributes=" + attributes
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/model/v1_0/UserDetails.java]: - [x] Re-gen

File: src/main/java/org/apereo/portlet/soffit/renderer/SoffitRendererController.java [hunk 12]
```diff
@@ -97,6 +104,14 @@ public ModelAndView render(final HttpServletRequest req, final HttpServletRespon
             // Deserialize the payload
             final Object soffit = objectMapper.readValue(soffitJson, payloadClass);
 
+            // Authorization header Bearer token
+            final String authorizationHeader = req.getHeader(Headers.AUTHORIZATION.getName());
+            final String bearerToken = authorizationHeader.substring(SoffitConnectorController.BEARER_PREFIX.length());
+            final UserDetails user = userDetailsService.parseBearerToken(bearerToken);
+
+System.out.println(" ## user="+user);
+System.out.println(" ## ");
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/renderer/SoffitRendererController.java]: - [x] Remove

File: src/main/java/org/apereo/portlet/soffit/renderer/SoffitRendererController.java [hunk 13]
```diff
@@ -97,6 +104,14 @@ public ModelAndView render(final HttpServletRequest req, final HttpServletRespon
             // Deserialize the payload
             final Object soffit = objectMapper.readValue(soffitJson, payloadClass);
 
+            // Authorization header Bearer token
```
Comment (drewwills) [inline src/main/java/org/apereo/portlet/soffit/renderer/SoffitRendererController.java]: - [x] Move to a method annotated with `@ModelAttribute`

**Comment by drewwills** [comment]
- [x] Objects with JWTs should implement a common interface
- [x] JWTs should include a Java class name that they deserialize to
- [x] Provide a `@Configuration` class that defines all the beans needed for Soffit
- [x] Map uPortal attributes to registered & public JWT claims wherever possible

**Comment by andrewstuart** [comment]
Any JWT implementation we do should definitely focus on mapping uPortal concepts to the [standardized JWT claims](https://tools.ietf.org/html/rfc7519#section-4.1). There's actually an [IANA registry](https://www.iana.org/assignments/jwt/jwt.xhtml) for JWT claims, so we should stick to that as closely as possible.

Groups/roles may be an exception, since nearly all the JWT registered claims are simple values rather than lists.

So basically, let's not implement JWT; there are at least 4 Java libraries listed on [jwt.io](//jwt.io#libraries-io). This probably means not including any java class name, although if this is important for some reason, you could stick it in the JWT header as a separate attribute. Either way, the Soffit backends should be able to reasonably choose one of the pre-existing JWT libraries to deserialize and still end up with usable claims (usually a map).

I don't have much time right this second, but wanted to make sure I got this out there as a start :smile:

**Pull request merged**